domain/machine/service/service.go

```diff
@@ -92,6 +102,12 @@ type State interface {
 
 	// IsMachineRebootRequired checks if the machine referenced by its UUID requires a reboot.
 	IsMachineRebootRequired(ctx context.Context, uuid string) (bool, error)
+
+	// GetMachineParentUUID returns the parent UUID of the specified machine.
+	// It returns a NotFound if the machine does not exist.
+	// It returns a MachineHasNoParent if the machine has no parent.
+	// It returns a GrandParentNotAllowed if the machine's parent has a parent.
+	GetMachineParentUUID(ctx context.Context, machineName machine.Name) (string, error)
```

Comment:
**todo:** I write the logic in https://github.com/juju/juju/pull/17755, but I will delete it (and related errors) and keep only the ShouldRebootOrRestart stuff.

However, after some thought, I don't think the check of grandparent is the responsibility of `GetMachineParentUUID`. 

From my perspective, this check should be done in the RebootOrShutdown function, since it would be this function that would requires updates if, for some reason, change on design make GrandParentMachine "a thing".

So it is what I have done in my own PR.

Reply:
> this check should be done in the RebootOrShutdown function

@gfouillet I disagree with this. Not allowing a grandparent is a design concern, not an internal workflow concern. It should be done either 1) at machine creation (basically anywhere that associate a machine with a parent), or 2) it would be at the GetMachineParent, as that's a service that handles the parental concern. I'd suggest going with option (2) right now just because it's there, but I think option (1) is better, so I can be convinced to change the code.

Also in a separate PR the RebootOrShutdown should be changed to use the GetMachineParent internally to get the parent, we have duplicate logic in two places right now.

Reply:
@cderici even though I initially made this same comment in https://github.com/juju/juju/pull/17755, I think I understand @gfouillet 's point: for sure we need to check at machine creation time (this is because of how we designed it) but checking it on reboot allows for checking at the actual hot-path (because that is the only place where we can actually fail if a machine has grandparents). 
And since https://github.com/juju/juju/pull/17755 is already merged, I propose to just add the check at machine creation.